server/src/test/java/io/crate/integrationtests/DynamicMappingUpdateITest.java

```diff
@@ -70,7 +70,7 @@ insert into t values (1, {x=1})
         HashMap<String, Integer> columnPositions = new HashMap<>();
         Thread checkPositions = new Thread(() -> {
             while (dmlStatementsFinished.get() == false) {
-                synchronized (response) {
+                synchronized (this) { // sqlExecutor and response objects are used by concurrent DML queries and DQL queries in the last assertions.
                     execute("select column_name, ordinal_position from information_schema.columns where table_name = 't'");
                     for (int i = 0; i < response.rowCount(); i++) {
```

Comment:
Won't the synchronized cancel out some of the concurrency we want in this test?

`execute` would return a `SQLResponse` that could be used instead to avoid the issue with concurrent reads and writes on the `response` attribute.

Reply:
> execute would return a SQLResponse that could be used instead

that was a previous version ` synchronized (response)` and it [failed](https://jenkins.crate.io/blue/organizations/jenkins/CrateDB%2Fcrate_test_hourly_stable/detail/crate_test_hourly_stable/10997/pipeline) in hourly tests as expected query result was from completely different query.

Reply:
The `synchronized (response)` is using the global `response` attribute. I'm suggesting to use a local variable for the result:


```diff
@@ -68,15 +69,15 @@ insert into t values (1, {x=1})

         AtomicBoolean dmlStatementsFinished = new AtomicBoolean(false);
         HashMap<String, Integer> columnPositions = new HashMap<>();
         Thread checkPositions = new Thread(() -> {
             while (dmlStatementsFinished.get() == false) {
-                synchronized (response) {
-                    execute("select column_name, ordinal_position from information_schema.columns where table_name = 't'");
-                    for (int i = 0; i < response.rowCount(); i++) {
-                        String columnName = (String) response.rows()[0][0];
-                        Integer newPosition = (Integer) response.rows()[0][1];
+                var response = execute("select column_name, ordinal_position from information_schema.columns where table_name = 't'");
+                for (int i = 0; i < response.rowCount(); i++) {
+                    String columnName = (String) response.rows()[0][0];
+                    Integer newPosition = (Integer) response.rows()[0][1];
+                    synchronized (columnPositions) {
                         Integer previousPosition = columnPositions.put(columnName, newPosition);
                         if (previousPosition != null && previousPosition.equals(newPosition) == false) {
                             throw new IllegalStateException(
                                 String.format(Locale.ENGLISH, "Column %s had position %d which is recomputed to %d", columnName, previousPosition, newPosition)
                             );
```

Reply:
Thx, adapted the suggestion with modifications:
I think I still need a block on `sqlExecutor` (otherwise can get result of another query like in https://jenkins.crate.io/blue/organizations/jenkins/CrateDB%2Fcrate_test_hourly_stable/detail/crate_test_hourly_stable/10997/pipeline/) and `synchronized (columnPositions)` is not needed as columnPositions is being written only by single thread.

Reply:
Maybe we can write to a temporary table and query the table later?

Reply:
> Maybe we can write to a temporary table and query the table later?

IIUC, we still need sqlExecutor sync since we want to get SQLResponse would be pretty much the same with a temporal table being used as a map?

Reply:
I meant something like
```
cr> create table pos (id int, column_name text, ordinal_position int);                                                                              
CREATE OK, 1 row affected  (1.628 sec)
cr> insert into pos select 1, column_name, ordinal_position from information_schema.columns;                                                        
INSERT OK, 956 rows affected  (0.180 sec)
```
and aggregate later on.

Reply:
> I think still need a block on sqlExecutor

It doesn't mutate any other state as far as I can tell?
If it did, each thread would need a synchronized.

Reply:
> this issue could be more than just column positions.

could you please post the whole error message? It could be a minor thing if it's only a test setup error if results on select-with-position (checker thread) gets messed up with select-without-positions. I [faced](https://jenkins.crate.io/blue/rest/organizations/jenkins/pipelines/CrateDB/pipelines/crate_test_hourly_stable/runs/10997/log/?start=0) that before and basically it was the reason to create this PR

Reply:
> add synchronized(sqlExecutor) to all DML threads. I wonder if this is actually a reason of position flakiness, maybe some [session/parsed statement](https://github.com/crate/crate/blob/master/server/src/testFixtures/java/io/crate/testing/SQLTransportExecutor.java#L305) of two DML-s gets mixed up?

The session is created _per_ `exec`. There is nothing shared across threads.

I think it's more likely that it surfaces issues with the implementation as it reduces contention in the test and increases actual concurrency.

Reply:
reproduced:
```
java.lang.AssertionError: 
Expecting actual:
  ["b",
    "b['newcol0']",
    "b['newcol1']",
    "b['newcol10']",
    "b['newcol11']",
    "b['newcol12']",
    "b['newcol13']",
    "b['newcol14']",
    "b['newcol2']",
    "b['newcol20']",
    "b['newcol21']",
    "b['newcol22']",
    "b['newcol23']",
    "b['newcol3']",
    "b['newcol30']",
    "b['newcol31']",
    "b['newcol32']",
    "b['newcol33']",
    "b['newcol34']",
    "b['newcol4']",
    "b['newcol40']",
    "b['newcol41']",
    "b['newcol42']",
    "b['newcol43']",
    "b['newcol44']",
    "b['newcol50']",
    "b['newcol51']",
    "b['newcol52']",
    "b['newcol53']",
    "b['newcol54']",
    "b['newcol60']",
    "b['newcol61']",
    "b['newcol62']",
    "b['newcol63']",
    "b['newcol64']",
    "b['newcol70']",
    "b['newcol71']",
    "b['newcol72']",
    "b['newcol73']",
    "b['newcol74']",
    "b['newcol80']",
    "b['newcol81']",
    "b['newcol82']",
    "b['newcol83']",
    "b['newcol84']",
    "b['x']"]
to contain exactly (and in same order):
  ["b",
    "b['newcol0']",
    "b['newcol1']",
    "b['newcol10']",
    "b['newcol11']",
    "b['newcol12']",
    "b['newcol13']",
    "b['newcol14']",
    "b['newcol2']",
    "b['newcol20']",
    "b['newcol21']",
    "b['newcol22']",
    "b['newcol23']",
    "b['newcol24']",
    "b['newcol3']",
    "b['newcol30']",
    "b['newcol31']",
    "b['newcol32']",
    "b['newcol33']",
    "b['newcol34']",
    "b['newcol4']",
    "b['newcol40']",
    "b['newcol41']",
    "b['newcol42']",
    "b['newcol43']",
    "b['newcol44']",
    "b['newcol50']",
    "b['newcol51']",
    "b['newcol52']",
    "b['newcol53']",
    "b['newcol54']",
    "b['newcol60']",
    "b['newcol61']",
    "b['newcol62']",
    "b['newcol63']",
    "b['newcol64']",
    "b['newcol70']",
    "b['newcol71']",
    "b['newcol72']",
    "b['newcol73']",
    "b['newcol74']",
    "b['newcol80']",
    "b['newcol81']",
    "b['newcol82']",
    "b['newcol83']",
    "b['newcol84']",
    "b['x']"]
but could not find the following elements:
  ["b['newcol24']"]
```